Java sprint 12 v1.0

1EVILGUN1 opened the pull request:
Добавлен пакет dao и inheritance.
в dao храняться прошлые storage классы переделанные под БД. Довалены классы Mpa и Genre и к ним же Service,Controller,DatabaseStorage. Добавлены обработки ошибок для Mpa, Genre, Film. Добавлено логгирование во всех классах контроллерах

1EVILGUN1 committed:
Java sprint 12 v1.0

Добавлен пакет dao и inheritance.
в dao храняться прошлые storage классы переделанные под БД. Довалены классы Mpa и Genre и к ним же Service,Controller,DatabaseStorage.
Добавлены обработки ошибок для Mpa, Genre, Film. Добавлено логгирование во всех классах контроллерах

1EVILGUN1 committed:
fix: delete target

1EVILGUN1 committed:
fix: DS_Store

1EVILGUN1 committed:
Update pom.xml

1EVILGUN1 committed:
fix:DS_Store

1EVILGUN1 committed:
fix: delete import ArrayList in FilmController

1EVILGUN1 committed:
fix: static SIZE_MPA, SIZE_GENRE

YazykovMikhail commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,76 @@
+
+CREATE TABLE IF NOT EXISTS users
+(
+    user_id  INTEGER               NOT NULL AUTO_INCREMENT,
```
просто id, если таблица назыввается users и там поле id, понятное дело что это его id
user_email у тебя же не назван и это верно)
Если бы тут была другая, на другю таблицу Заказ например, тогда да request_id, а так, если поля относяться к этой таблице/классу приставки там не нужны

YazykovMikhail replied:
Поправь везде

YazykovMikhail commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,76 @@
+    description  CHARACTER VARYING(200) NOT NULL,
+    release_date DATE                   NOT NULL,
+    duration     INTEGER                NOT NULL,
+    rating_id    INTEGER                NOT NULL,
```
вот тут верно, это ссылка на другую табоицу, тут и должно быть rating_id

YazykovMikhail commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,76 @@
+        ON DELETE CASCADE ON UPDATE CASCADE,
+    CONSTRAINT friendship_fk_1 FOREIGN KEY (user_second_id) REFERENCES users (user_id)
+        ON DELETE CASCADE ON UPDATE CASCADE
+);
```
Можно так же сделать эти два поля состаным ключом:
PRIMARY KEY (user_first_id, user_second_id)

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDatabaseStorage.java:
```diff
@@ -0,0 +1,190 @@
+@RequiredArgsConstructor
+public class FilmDatabaseStorage implements FilmStorage {
+    private final JdbcTemplate jdbcTemplate;
+    private final UserStorage userStorage;
```
Стораджи между собой не должны быть зависимы, все необходимые данные должны приходить из сервиса, если нужно какие то данные из другого стораджа, они должны вызываться из сервиса и уже готовые данные приходить сюда. Каждый сторадж отвечает только за свою модель

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDatabaseStorage.java:
```diff
@@ -0,0 +1,190 @@
+
+    @Override
+    public Collection<Film> getFilms() {
+        Map<Long, Film> films = new HashMap<>();
```
а зачем эта Мапа, ты создаешь пустую мапу, получаешь список фильмов, из списка делаешь мапу и потом обратно, из мапы список

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -17,44 +20,51 @@
 
-    public List<Film> getFilms() {
+    public Collection<Film> getFilms() {
         return filmStorage.getFilms();
```
С Жанрами связь мнгие ко многим, нужно делать умный запрос, что бы не вызывать для каждого фильмы свои жанры, ведь делать множетсво запросов для списка - это плохо, т,к для каждого создается соединение, это тяжеловестная операция, стоит делать умный запрос к бд, что бы он был 1.

Поэтому в GenreStorage стоит сделать метод:
```
	@Override
	public void load(List<Film> films) {
		String inSql = String.join(",", Collections.nCopies(films.size(), "?"));
		final Map<Long, Film> filmById = films.stream().collect(Collectors.toMap(Film::getId, identity()));
		final String sqlQuery = "select * from GENRES g, FILM_GENRES fg where fg.GENRE_ID = g.GENRE_ID AND fg.FILM_ID in (" + inSql + ")";
		jdbcTemplate.query(sqlQuery, (rs) -> {
			final Film film = filmById.get(rs.getLong("FILM_ID"));
			film.addGenre(makeGenre(rs, 0));
		}, films.stream().map(StorageData::getId).toArray());
	}

	static Genre makeGenre(ResultSet rs, int rowNum) throws SQLException {
		return new Genre(
				rs.getLong("genre_id"),
				rs.getString("name"));
	}
	
```
Который будет возвращать данные толкьо для необходимых фильмов и наполнять фильмы нужными жанрами.
и соответсвенно тут еще вызывать этот метод:
```
	@Override
	public List<Film> getFilms() {
		final List<Film> films = filmStorage. getFilms();
		genreStorage.load(films);
		return films;
	}
```

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDatabaseStorage.java:
```diff
@@ -0,0 +1,190 @@
+                .build();
+        log.info("Film build end>>>>>");
+        log.info("Film = {}", film);
+        List<Genre> genresOfFilm = getGenresOfFilm(film.getId());
```
Вот так делать не надо, получается когда получаются все фильмы, они в цикле вызывают дополнительные запросы на каждый фильм, что бы заполинть жанры, так лучше не делать. опсиал в сервисе как стоит сделать

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDatabaseStorage.java:
```diff
@@ -0,0 +1,190 @@
+        log.info("Film build end>>>>>");
+        log.info("Film = {}", film);
+        List<Genre> genresOfFilm = getGenresOfFilm(film.getId());
+        List<Integer> likes = getLikesOfFilm(film.getId());
```
Лайки в фильмах сейчас вообще не нуны, раньше они были тут для хранения, и на них даже была анноатция @JsonIgnore, а сейчас для их ранения есть отдельная БД и работа с ними через отдельные контролелры

1EVILGUN1 committed:
fix: перенос из filmstorage

Добавлен likestorage и перенесены методы из filmstorage. Также относиться к mpa,genre storage
